teps/0071-custom-task-sdk.md

```diff
@@ -0,0 +1,220 @@
+last-updated: '2021-06-15'
+authors:
+- '@ScrapCodes'
+---
+
+# TEP-0071: Custom Task SDK
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+    - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+    - [Notes/Caveats (optional)](#notescaveats-optional)
+    - [Risks and Mitigations](#risks-and-mitigations)
+    - [User Experience (optional)](#user-experience-optional)
+    - [Performance (optional)](#performance-optional)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [Implementation Pull request(s)](#implementation-pull-request-s)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+A Custom Task SDK will make the work of custom task author easier and will offer
+standard way to implement custom task controllers. While adhering to current
+separation of concern between a custom task controller's role and tektoncd owned
+controller's role, it should make custom tasks easier to manage and reason by
+tektoncd.
+
+## Motivation
+
+1. Currently, a custom task author has limited ways to know what is a standard way
+    of building his custom task. His best bet is to explore one of the existing 
+   implementation in `experimental/` repo.
+2. However, an experimental custom task may not be up to date with the latest `tektoncd`
+   code base version. There is no one goto place, where a Task author can find upto
+   date example.
+3. Presently, `tektoncd` does not manage the life cycle of a custom task at all.
+   A SDK can open up newer possibilities in this regard.
+4. SDK available as a project template, can perform some of common tasks,
+   generate boiler-plate code and provide helper functions/scripts.
+
+### Goals
+
+1. Provide a Github template with a simple controller and CRD implementation,
+   which a user can fork and develop his own custom task controller.
+2. Provide helper functions/packages for common tasks and documentation
+   with best practices.
+
+### Non-Goals
+
+TBD.
+
+### Use Cases (optional)
+
+<!--
+Describe the concrete improvement specific groups of users will see if the
+Motivations in this doc result in a fix or feature.
+
+Consider both the user's role (are they a Task author? Catalog Task user?
+Cluster Admin? etc...) and experience (what workflows or actions are enhanced
+if this problem is solved?).
+-->
+
+## Requirements
+
+<!--
+Describe constraints on the solution that must be met. Examples might include
+performance characteristics that must be met, specific edge cases that must
+be handled, or user scenarios that will be affected and must be accomodated.
+-->
```

Comment:
ideas:

- A user should be able to fork the SDK and and run it with no modifications required (e.g. it provides a working sample out of the box)
- We version the SDK (i think? or maybe not?)
- We could list everything the sample should include, e.g.: working controller, example configuration (built with ko?), docs, tests (unit and integration maybe? maybe system tests too)

It also might be worth describing in detail what the behavior of the example custom task would be (but np if we want to add that as we go)

Reply:
@bobcatfish these are great ideas, I was thinking, maybe I can add them in the proposal step. 

Thank you very much for taking a look.